src/openrct2/ride/Track.cpp

```diff
@@ -899,3 +899,104 @@ void TrackElement::SetHighlight(bool on)
     if (on)
         Flags2 |= TRACK_ELEMENT_FLAGS2_HIGHLIGHT;
 }
+
+bool TrackSupportsHasFlag(uint8_t queryFlag, uint8_t sequenceFlags)
+{
+    switch (queryFlag)
+    {
+        case TRACK_SUPPORT_HAS_SUPPORT:
+        case TRACK_SUPPORT_RCT1_EXTRA_SUPPORT:
+        case TRACK_SUPPORT_INVERSION_EXTRA_SUPPORT:
+        case TRACK_SUPPORT_COVERED_EXTRA_SUPPORT:
+        case TRACK_SUPPORT_SHOULD_HAVE_SUPPORT:
+            return sequenceFlags & queryFlag;
+        case TRACK_SUPPORT_LOOP_A:
+        case TRACK_SUPPORT_LOOP_B:
+        case TRACK_SUPPORT_LOOP_C:
+            return (sequenceFlags & TRACK_SUPPORT_CHAIRLIFT_SUPPORT) != TRACK_SUPPORT_CHAIRLIFT_SUPPORT
+                && sequenceFlags & queryFlag;
+        case TRACK_SUPPORT_CHAIRLIFT_SUPPORT:
+            return (sequenceFlags & TRACK_SUPPORT_CHAIRLIFT_SUPPORT) == TRACK_SUPPORT_CHAIRLIFT_SUPPORT;
+        default:
+            return false;
+    }
+}
+
+bool TrackSupportDefaultState(RideTypeDescriptor rtd, CoordsXY position, track_type_t trackType, uint8_t sequence)
+{
+    if (rtd.SupportsBehaviour & RideSupportsBehaviour::AlwaysHasSupports)
+    {
+        return true;
+    }
+    else
+    {
+        Guard::Assert(sequence < MaxSequencesPerPiece);
+        auto ted = GetTrackElementDescriptor(trackType);
+        // if it is only a chairlift, only give chairlift supports
+        if (rtd.SupportsBehaviour == RideSupportsBehaviour::ChairliftSupports)
+        {
+            return TrackSupportsHasFlag(TRACK_SUPPORT_CHAIRLIFT_SUPPORT, sequence);
+        }
+        bool shouldDraw = shouldDraw = TrackSupportsHasFlag(TRACK_SUPPORT_HAS_SUPPORT, ted.SequenceSupportBehaviour[sequence]);
+        if (TrackSupportsHasFlag(TRACK_SUPPORT_SHOULD_HAVE_SUPPORT, ted.SequenceSupportBehaviour[sequence]))
+        {
+            if (rtd.SupportsBehaviour & RideSupportsBehaviour::ShouldHaveSupports)
+                shouldDraw |= ((position.x & (1 << 5)) == (position.y & (1 << 5)))
+                    || ((!(position.x & (1 << 5))) && (!(position.y & (1 << 5))));
+            else
+                shouldDraw = true;
+        }
+
+        if (rtd.SupportsBehaviour & RideSupportsBehaviour::ChairliftSupports)
+            shouldDraw |= TrackSupportsHasFlag(TRACK_SUPPORT_CHAIRLIFT_SUPPORT, ted.SequenceSupportBehaviour[sequence]);
+
+        if (rtd.SupportsBehaviour & RideSupportsBehaviour::RCT1ExtraSupports)
+            shouldDraw |= TrackSupportsHasFlag(TRACK_SUPPORT_RCT1_EXTRA_SUPPORT, ted.SequenceSupportBehaviour[sequence]);
+
+        if (rtd.SupportsBehaviour & RideSupportsBehaviour::CoveredExtraSupports)
+            shouldDraw |= TrackSupportsHasFlag(TRACK_SUPPORT_COVERED_EXTRA_SUPPORT, ted.SequenceSupportBehaviour[sequence]);
+
+        if (rtd.SupportsBehaviour & RideSupportsBehaviour::InversionExtraSupports)
+            shouldDraw |= TrackSupportsHasFlag(TRACK_SUPPORT_INVERSION_EXTRA_SUPPORT, ted.SequenceSupportBehaviour[sequence]);
```

Comment:
Inverting coasters tend to have 2 supports on inline twists/barrel rolls, while non-inverted coasters tend to have only 1 support